src/main/scala/org/apache/spark/sql/catalyst/plans/logical/merge.scala

```diff
@@ -244,7 +244,7 @@ object MergeInto {
       resolvedExpr.flatMap(_.references).filter(!_.resolved).foreach { a =>
         // Note: This will throw error only on unresolved attribute issues,
         // not other resolution errors like mismatched data types.
-        val cols = "columns " + plan.references.map(_.sql).mkString(", ")
+        val cols = "columns " + plan.children.flatMap(_.output).map(_.sql).mkString(", ")
```

Comment:
printing all the `plan.references` also prints invalid column names which should not be printed here. instead it should only the print the valid reference which are the output attributes of the children of this plan.